[Pytorch Delegated Backend] Save function name in debug info

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58873 [Pytorch backend delegation] Preprocess to accept
* **#57481 [Pytorch Delegated Backend] Save function name in debug info**
* #57441 [Pytorch Delegated Backend] Save operator name and function name in

Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

kimishpatel committed:
[Pytorch Delegated Backend] Save function name in debug info

Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

ZolotukhinM commented on torch/csrc/jit/ir/scope.h:
```diff
@@ -120,6 +120,15 @@ struct TORCH_API InlinedCallStack : public c10::intrusive_ptr_target {
   c10::optional<InlinedCallStackPtr> callee_;
   Function* fn_;
+  // Reason for fn_name_ even though we have fn_
+  // Serialized callstack is used in circustmances where InlinedCallstack
```
Typo: circumstances

ZolotukhinM reviewed (approved):
Looks good to me!

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

raziel commented on torch/csrc/jit/serialization/callstack_debug_info_serialization.cpp:
```diff
@@ -40,6 +46,16 @@ c10::IValue InlinedCallStackSerializer::serialize(
+    elements.emplace_back(cs_ptr->function()->name());
+  } else {
+    auto fn_name = cs_ptr->function_name();
+    if (!fn_name.empty()) {
```
Curious, aren't we already making sure that this is always populated (with FunctionName_UNKNOWN in the worst case)?

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).

kimishpatel committed:
Update on "[Pytorch Delegated Backend] Save function name in debug info"


Summary:
This diff introduces function name to InlinedCallStack.
Since we are using InlinedCallStack for debug information in lite
interpreter as well as delegate backends, where InlinedCallStack cannot
be constructed from model source code, we need to save function name.
In the absence of function name Function* is used to get name of the
function. This is when JIT compiles code at runtime.
When that is not possible, this diff introduces a way to obtain function
name.

Test Plan:
test_backend
test_cs_debug_info_serialization

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28159097](https://our.internmc.facebook.com/intern/diff/D28159097)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28159097).
